### ISSUE TITLE
Excessive Logging from CronJob

### ISSUE THREAD

**Issue by daviddenis-stx** [kubernetes/kubernetes#118789]
### What happened?

Paraphrasing this https://github.com/canonical/microk8s/issues/3908

concurrencyPolicy:Forbid and scheduling interval smaller than job actual runtime leads to lots of events sent by control plane (ending up clogging logs on observability side of things)

For example for a job with a once per 55 minutes schedule and actual execution being between a few minutes (nothing much to do) an a few hours (lots of things to do) the control plane sent ~16E6 log lines in 25 minutes.

Also, tenth of log lines are seen within the same millisecond, it smells like a quite active loop producing the same event as fast as it can over and over again.

<img width="1589" alt="image" src="https://github.com/kubernetes/kubernetes/assets/50578466/9b94b6d8-37c7-4089-8750-1b038ff6451e">

### What did you expect to happen?

Probably much less events (or logging) ?

### How can we reproduce it (as minimally and precisely as possible)?

- Create a CronJob that runs for a long time with concurrencyPolicy: Forbid
- Have the second attempt to run the job start while the previous one is still running

### Kubernetes version

<details>

```yaml
clientVersion:
  buildDate: "2023-05-17T14:20:07Z"
  compiler: gc
  gitCommit: 7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647
  gitTreeState: clean
  gitVersion: v1.27.2
  goVersion: go1.20.4
  major: "1"
  minor: "27"
  platform: linux/amd64
kustomizeVersion: v5.0.1
serverVersion:
  buildDate: "2022-12-08T10:08:09Z"
  compiler: gc
  gitCommit: 804d6167111f6858541cef440ccc53887fbbc96a
  gitTreeState: clean
  gitVersion: v1.25.5
  goVersion: go1.19.4
  major: "1"
  minor: "25"
  platform: linux/amd64
```

</details>

### Cloud provider

On premises

### OS version

<details>

PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

containerd containerd.io 1.6.21 3dce8eb055cbb6872793272b4f20ed16117344f8

</details>

**Comment by aramhakobyan** [comment]
Hello, @xuexu6666, 
 we have hit the issue heavily already several times. 
 Do you perhaps have some ETA on when the bug could be solved?

**Comment by gerasym** [comment]
I got another reproduce of the bug, but with different log message. Still likely also caused by the same race/infinite loop.
https://github.com/kubernetes/kubernetes/blob/v1.25.14/pkg/controller/cronjob/utils.go#L96-L97C23 we got >1.3 billions of this message generated by 3 cronjobs with * * * * * schedule. Masters were running at 100% CPU and filling up the disk with logs. Eventually it brought the cluster to read only state. As soon as you delete the affected cronjobs the issue stops.

**Comment by kannon92** [comment]
Thinking about this, kubelet should be rotating these logs to avoid having them fill up.  Could you search the logs in kubelet for rotating logs for KCM (Controller manager).

There is a PR to do the log rotations in parallel when there are lots of files.

**Comment by gerasym** [comment]
It doesn't seem it logs anything when it is successful - https://github.com/kubernetes/kubernetes/blob/v1.25.14/pkg/kubelet/logs/container_log_manager.go#L413

Also not sure if rotation was happening? I don't see it in kubelet args and i don't see in in kubelet config file, defaults are 10Mb and 5 files. Maybe they were not used for some reason?

**Comment by insertjokehere** [comment]
I am also running into this issue, running k3s on-prem. If a cronjob with concurrencyPolicy set to Forbid triggers while a previous instance is still running, `/var/log/syslog` fills up with `Not starting job because prior execution is running and concurrency policy is Forbid`, which eventually causes disk pressure on all nodes

**Comment by kannon92** [comment]
@atiratree can you help triage this?

**Comment by atiratree** [comment]
having a backoff for these events would certainly be beneficial

for now a workaround is to make sure that the interval between schedules is always less than the duration of the job 

/triage accepted
/priority important-longterm

**Comment by flavianmissi** [comment]
Tried the current assignee both here and on k8s slack but got no reply. I'm assuming there isn't any work happening on this so I'll take over.
/assign

**Comment by flavianmissi** [comment]
I'm trying to reproduce this on a local kind cluster without much luck. 
This is what did (exact steps down below):

1. create a cronjob that runs a job every minute
2. the job in question sleeps for an hour

And I do not see a flood of `JobAlreadyActive` events. The number of events averages 1.5 event per minute... nowhere near the ~16E6 in 25 minutes experienced by the reporter 🤔
I wonder if I missed a step? Below is what I tried.

```
$ cat cronjob.yaml
apiVersion: batch/v1
kind: CronJob
metadata:
  name: testcron
  namespace: test
spec:
  schedule: '* * * * *'
  concurrencyPolicy: Forbid
  jobTemplate:
    metadata:
      creationTimestamp: null
      labels:
    spec:
      template:
        spec:
          containers:
          - command:
            - /bin/sh
            - -c
            - sleep 3600; exit 0
            image: docker.io/library/busybox
            imagePullPolicy: Always
            name: my-job
            resources: {}
            terminationMessagePath: /dev/termination-log
            terminationMessagePolicy: File
          dnsPolicy: ClusterFirst
          restartPolicy: Never
          schedulerName: default-scheduler
          securityContext: {}
          terminationGracePeriodSeconds: 30
  successfulJobsHistoryLimit: 3
  failedJobsHistoryLimit: 1
$ kubectl apply -f cronjob.yaml

# leave it running for several minutes

$ oc describe cronjob/testcron # edited for readability
Name:                          testcron
Namespace:                     test
[...]
Last Schedule Time:  Mon, 17 Jun 2024 17:44:00 +0200
Active Jobs:         testcron-28643984
Events:
  Type    Reason            Age                    From                Message
  ----    ------            ----                   ----                -------
  Normal  JobAlreadyActive  105s (x171 over 116m)  cronjob-controller  Not starting job because prior execution is running and concurrency policy is Forbid
$ oc logs kube-controller-manager-kind-control-plane -nkube-system | grep JobAlreadyActive | wc -l
     172
```
What am I missing?

**Comment by alculquicondor** [comment]
Did you try it with k8s 1.25? Or a different version?

**Comment by flavianmissi** [comment]
Good thinking, my previous test was on v1.29. 
Testing with v1.25.16 yield very similarly reasonable results:
```
$ kubectl describe cronjob/testcron # edited for readability
Name:                          testcron
Namespace:                     test
Events:
  Type    Reason            Age                 From                Message
  ----    ------            ----                ----                -------
  Normal  SuccessfulCreate  36m                 cronjob-controller  Created job testcron-28646442
  Normal  JobAlreadyActive  48s (x36 over 35m)  cronjob-controller  Not starting job because prior execution is running and concurrency policy is Forbid

$ kubectl logs kube-controller-manager-kind-control-plane -nkube-system | grep JobAlreadyActive | wc -l
     38
```
If I had to guess only by looking at the code I'd say that the call to [`nextScheduleTimeDuration`](https://github.com/openshift/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controllerv2.go?plain=1#L570) is miscalculating the requeue time, causing too many requeues too fast. 

Looking at the introspection report uploaded to https://github.com/canonical/microk8s/issues/3908 I can see the log occurrences happen a few thousand times within a single second.
```
$ grep -E 'I0408 03:57:00.*  839288 event.go:294] "Event occurred" object="default/cron-sitemap" fieldPath="" kind="CronJob" apiVersion="batch/v1" type="Normal" reason="JobAlreadyActive" message="Not starting job because prior execution is running and concurrency policy is Forbid"' snap.microk8s.daemon-kubelite/journal.log | wc -l
3961
```
The cronjob controller is definitely unhinged, but it doesn't seem like the scheduling combined with concurrencyPolicy:Forbid is the sole cause of the issue. It could help to look the entire cronjob yaml that is causing the issue (spec + status). If anyone here has it please do share, I'll ask on the microk8s issue as well, maybe I get lucky.

**Comment by insertjokehere** [comment]
Here is the full YAML for one of my jobs that was triggering this issue straight out of `kubectl get -o yaml`, sorry its a bit messy. I'm working around it by now by managing a lock file manually and have updated the cluster a couple of times since this was last an issue.

fwiw, I'm currently running v1.27.2+k3s1 (v1.27.2, via k3s)

[backup.yml.txt](https://github.com/user-attachments/files/15898794/backup.yml.txt)

**Comment by flavianmissi** [comment]
@insertjokehere if it's not too much to ask, could you run the cronjob below in your k3s cluster and see if you can reproduce the problem with it? 
If you are able to do this, could you also share the cronjob yaml (specially the status) from when the problem is being observed? In other words, I'd like to see the cronjob status while the `JobAlreadyActive` event is crazy looping. Additionally the logs of the controller manager would also be extremely helpful, preferably with log level 4 (so we see the exact value for [`requeueAfter`](https://github.com/openshift/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controllerv2.go?plain=1#L219)).

cronjob.yaml:
```yaml
apiVersion: batch/v1
kind: CronJob
metadata:
  name: testcron
  namespace: test
spec:
  schedule: '* * * * *'
  concurrencyPolicy: Forbid
  jobTemplate:
    metadata:
      creationTimestamp: null
      labels:
    spec:
      template:
        spec:
          containers:
          - command:
            - /bin/sh
            - -c
            - sleep 3600; exit 0
            image: docker.io/library/busybox
            imagePullPolicy: Always
            name: my-job
            resources: {}
            terminationMessagePath: /dev/termination-log
            terminationMessagePolicy: File
          dnsPolicy: ClusterFirst
          restartPolicy: Never
          schedulerName: default-scheduler
          securityContext: {}
          terminationGracePeriodSeconds: 30
  successfulJobsHistoryLimit: 3
  failedJobsHistoryLimit: 1
```

For some background: I tried reproducing this on kind versions 1.29 and 1.25, as well as OpenShift 4.12 (based k8s 1.25). On the OpenShift cluster I also set some skew to the clock of two (out of three) control plane nodes, both to the past and the future, and neither helped me reproduce this issue. Without any more data to analyze, all I can think to do is play with [unit tests](https://github.com/openshift/kubernetes/blob/v1.25.9/pkg/controller/cronjob/utils_test.go#L280) and see if I can force some strange situation there (🤞🏼), so this is what I'll be working on until I can get more data.

Many thanks!

**Comment by flavianmissi** [comment]
I was able to somewhat reproduce this issue. After putting my kind cluster to sleep for the weekend (by shutting down the docker qemu vm), upon turning the cluster back on I saw thousands of `TooManyMissedTimes`, just like @gerasym.
The good news is that this problem doesn't happen on master. Long story short, this happens because [`requeueAfter`](https://github.com/openshift/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controllerv2.go#L218-L221) is negative. This is the line of code that stops that from happening: https://github.com/openshift/kubernetes/blame/master/pkg/controller/cronjob/utils.go#L201.
I also able to show this behavior on a test case and a on [Go's playground](https://go.dev/play/p/FiqXwGd1bXf). I'll see if I can send a PR in with the test case so that we can avoid potential regressions in the future.

I'm still chasing the original `JobAlreadyActive` problem.

**Comment by insertjokehere** [comment]
> if it's not too much to ask, could you run the cronjob below in your k3s cluster and see if you can reproduce the problem with it?

Hi - yeah, absolutely can; busy week last week but I should have some time this weekend.

When you say

> Additionally the logs of the controller manager would also be extremely helpful, preferably with log level 4

How do I configure that?

Thanks for looking into this

**Comment by flavianmissi** [comment]
> How do I configure that?

According to [this comment](https://github.com/kubernetes/kubernetes/issues/22723#issuecomment-194034919), this can be done by editing the static pod manifest for kube-controller-manager, setting the `-v` flag to >= 4 in the pod command.
Alternatively, [this blog post](https://nakamasato.medium.com/how-to-change-log-level-of-kubernetes-components-3cd0e15b86ea) states this can be done at runtime with a few setup steps.

**Comment by flavianmissi** [comment]
I was able to reproduce the main problem of too many `JobAlreadyActive` events on 1.25.5, although I'm not exactly sure how (a combination of letting the job run for multiple days and stopping/starting my kind cluster maybe?).

Details below.

---
TL;DR:
I haven't tested on every z-stream, but I could only reproduce it on v1.25.5-v1.25.11 (I did not test it on versions earlier than v1.25.5) - it seems the bug is fixed on v1.25.11-v1.25.16, as well as on v1.26.15, v1.27.14, v1.28.10, and v1.29.5.

In short, I cannot reproduce this on master. There might be a chance that with the right combination of cron schedule, `.spec.lastScheduleTime` and `time.Now()` the bug happens again, in which case we will need these values to reproduce it (and thus fix it).

The outcome of this investigation is to send a PR with unit tests reproducing both the [`TooManyMissed`](https://github.com/kubernetes/kubernetes/issues/118789#issuecomment-1790501523) and `JobAlreadyActive` issues reported in this bug. They will pass on master but it'll prevent us from regressing. I'll work on that in the next few days.

---

Now for the long explanation 😄 

Here's the event part of `kubectl describe cj <cronjob>`:
```
Events:
  Type    Reason            Age                      From                Message
  ----    ------            ----                     ----                -------
  Normal  JobAlreadyActive  14s (x4670618 over 79m)  cronjob-controller  Not starting job because prior execution is running and concurrency policy is Forbid
```
4670618 times over 79m is nearly 6 thousand times per minute!

Here are the relevant parts of my cronjob yaml:
```yaml
apiVersion: batch/v1
kind: CronJob
# [...]
metadata:
  creationTimestamp: "2024-06-19T13:19:27Z"
  # [...]
spec:
  concurrencyPolicy: Forbid
  # [...]
  schedule: 0/55 * * * *
  successfulJobsHistoryLimit: 3
  suspend: false
status:
  active:
  - apiVersion: batch/v1
    kind: Job
    name: testcron2-28655515
    namespace: test
    # [...]
  lastScheduleTime: "2024-06-25T15:55:00Z"
  lastSuccessfulTime: "2024-06-25T15:57:37Z"
```

The following test case, written using the parameters from the job above, returns a negative duration on v1.25.5:
```go
func TestNextScheduleTimeDurationAfterSuspend(t *testing.T) {
       creationTime, err := time.Parse(time.RFC3339, "2024-06-19T13:19:27Z")
       if err != nil {
               t.Fatal(err)
       }
       earliestTime, err := time.Parse(time.RFC3339, "2024-06-25T15:55:00Z")
       if err != nil {
               t.Fatal(err)
       }
       metav1EarliestTime := metav1.NewTime(earliestTime)
       now, err := time.Parse(time.RFC3339, "2024-06-26T15:19:42Z")
       if err != nil {
               t.Fatal(err)
       }

       cj := batchv1.CronJob{
               ObjectMeta: metav1.ObjectMeta{
                       CreationTimestamp: metav1.NewTime(creationTime),
               },
               Spec: batchv1.CronJobSpec{
                       Schedule: "0/55 * * * *",
               },
               Status: batchv1.CronJobStatus{
                       LastScheduleTime: &metav1EarliestTime,
               },
       }

       expectedDuration := 27*time.Hour + 100*time.Millisecond

       sched, err := cron.ParseStandard(cj.Spec.Schedule)
       if err != nil {
               t.Errorf("error setting up the test, %s", err)
       }
       gotScheduleTimeDuration := nextScheduledTimeDuration(cj, sched, now)
       if *gotScheduleTimeDuration < 0 {
               t.Errorf("scheduleTimeDuration should never be less than 0, got %s", gotScheduleTimeDuration)
       }
       if !reflect.DeepEqual(gotScheduleTimeDuration, expectedDuration) {
               t.Errorf("scheduleTimeDuration - got %s, want %s", gotScheduleTimeDuration, expectedDuration)
       }
}
```

Running that on branch v1.25.5 (note that the logs you see were added by me to aid debugging):
```
$ go test -v -run TestNextScheduleTimeDurationAfterSuspend ./pkg/controller/cronjob
=== RUN   TestNextScheduleTimeDurationAfterSuspend
I0626 17:36:24.803233   98814 cronjob_controllerv2.go:660] "nextScheduledTimeDuration: earliest time is last schedule time" cronjob="" earliestTime="2024-06-25 15:55:00 +0000 UTC"
I0626 17:36:24.803363   98814 utils.go:120] "getMostRecentScheduleTime: params" earliestTime="2024-06-25 15:55:00 +0000 UTC" now="2024-06-26 15:19:42 +0000 UTC"
I0626 17:36:24.803375   98814 utils.go:123] "getMostRecentScheduleTime: t1,t2" t1="2024-06-25 16:00:00 +0000 UTC" t2="2024-06-25 16:55:00 +0000 UTC"
I0626 17:36:24.803381   98814 utils.go:139] "getMostRecentScheduleTime:" timeBetweenTwoSchedules=3300
I0626 17:36:24.803385   98814 utils.go:141] "getMostRecentScheduleTime:" timeElapsed=83982
I0626 17:36:24.803389   98814 utils.go:143] "getMostRecentScheduleTime:" numberOfMissedSchedules=26
I0626 17:36:24.803396   98814 utils.go:145] "getMostRecentScheduleTime: returning t (most recent time)" t="2024-06-26 14:55:00 +0000 UTC"
I0626 17:36:24.803405   98814 cronjob_controllerv2.go:674] "nextScheduledTimeDuration: returning t" cronjob="" t="-19m41.9s"
    utils_test.go:415: scheduleTimeDuration should never be less than 0, got -19m41.9s
    utils_test.go:418: scheduleTimeDuration - got -19m41.9s, want 27h0m0.1s
--- FAIL: TestNextScheduleTimeDurationAfterSuspend (0.00s)
FAIL
FAIL	k8s.io/kubernetes/pkg/controller/cronjob	0.960s
FAIL
```
Also note that `27h0m0.1s` (the expected duration of the test) is a guess and not the actual expected value.

I also run this same test on branches v1.29.5, v1.28.10, v1.27.14 and v1.26.15 and all consistently return `35m18.1s`.
In addition to those, I also ran the test on v1.25.5 (`-19m41.9s`) through v1.25.16 (`35m18.1s`), and I can reproduce the bug from v1.25.5 all the way through v1.25.11. From v1.25.12 until v1.25.16 the test consistently returns `35m18.1s`.

**Comment by flavianmissi** [comment]
While writing a unit test case to cover my reproducers I noticed that [an existing test case](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/utils_test.go#L647-L661) already covers a similar situation (where [`nextScheduleTimeDuration`](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/utils.go#L190) might return a negative value) - I brought the entire `TestNextScheduleTimeDuration` suite back to 1.25 and `another complex schedule skipping weekend` test fails there.
Unless we observe this problem on kube >= v1.26 I don't think we need any new tests or a code fix.
In other words, I think we can close this as resolved. 
And if anyone experiences anything like this again please bring us:
1. the cronjob yaml (including the status)
2. kube-controller-manager logs showing the excessive logs
3. it's important that 1 and 2 be captured around the same time, do not edit the cronjob until you capture the requested info or we won't be able reproduce the problem

---
An interest side quest could be to revisit the behavior of [`queue.AddAfter`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go?plain=1#L241-L245) - right now it treats negative numbers as "no delay". I'm not sure if there's an interest to change this behavior, let me know what y'all think.

**Comment by alculquicondor** [comment]
Actually, 1.28 is the oldest supported version.

Thanks @flavianmissi for the thorough investigation!

/close
